### DISCUSSION TITLE
Poisson Solve

### DISCUSSION THREAD

**Discussion by alahyari** [AMReX-Codes/amrex#2826] [Q&A]
Hello AMReX folks,

I have written this 3D vorticity-velocity-based solver using AMReX. The main solution contains 3 vorticity and 3 velocity components that are all cell-centered. We have 3 vorticity transport equations (one in each direction) that updates the vorticity field from time step n to n+1 based on the velocity field at time step n, and three 3 Poisson equations (one for each velocity components) that give the velocity field at time step n+1 where the RHS of the Poisson equation is a function of vorticity (at time step n+1).

With this brief introduction, here is my question: I was trying to solve a 2D x-y problem (convecting Taylor vortex) using this 3D code, with only one cell in the neutral direction, z (I tried more cells and it didn't help). The BC is periodic in all direction. The problem is when I use the built-in AMReX periodicity, everything is good and the Poisson solve seems to be working in all directions just fine (I check that against the exact solution), but when I enforce periodicity using my external (user-defined) BC function with Dirichlet BC for the linear solver, the Poisson solve returns wrong solution. This is odd, because when I compare the RHS values as well as the solution value of the ghost cells for both AMReX periodic BCs and the external BC function right before calling the solve, they are identical, but the solutions are different! 

Has anyone encountered a similar problem? Any advice would be greatly appreciated!

Thank you

**Comment by WeiqunZhang** [comment]
>  but when I enforce periodicity using my external (user-defined) BC function with Dirichlet BC for the linear solver,

Could you explain what that means?

**Reply by alahyari** [reply]
Of course. I mean instead of using AMReX periodic BC in all direction, I set the BCRec to ext_dir in all directions, and use the external function to enforce periodicity (i.e., U[-1] = U[N-1], U[N] = U[0], etc.). Further the BC for the linear solver is set to LinOpBCType::Dirichlet. Is that clear?

**Reply by drummerdoc** [reply]
This sounds like maybe there's a misunderstanding.  Typically, INT_DIR is used to indicate a periodic boundary, and geom.is_periodic is set to a zero or one in each direction as well.  Are you doing something special where this "usual" way is not applicable?

**Reply by WeiqunZhang** [reply]
Unless you know the solution already, you will not be able to use Dirichlet boundaries to emulate periodic boundaries in the linear solver.  You  should be able to do that in an explicit advance step if you are careful.  That is U[-1] = U[N-1] needs to involve parallel communication if necessary.

**Reply by WeiqunZhang** [reply]
More information. For the linear solver, one must call `setLevelBC`.  If it has Dirichlet BCs, one must provide the Dirichlet values in the function argument.

https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLLinOp.H#L164

**Reply by alahyari** [reply]
Thank you both for the quick replies. Yes, I should have mentioned that, I already know the solution everywhere, there is an analytical solution to the problem. And I'm running a serial test case (i.e., no parallel communication needed). So, I have the solution everywhere (i.e., U[0] and U[N-1] are known), then I set U[-1] = U[N-1], U[N] = U[0], and use setLevelBC to set Dirichlet BCs for the linear solver.

Just to provide a little bit more information, as explained above, I'm trying to simulate a 2D problem on a 3D code, so I have a neutral direction. The 2D problem in x-y plane is periodic, but in the z direction (the neutral direction), in theory you can use either periodic or symmetry or slip wall BC. When I use periodic BC where geom.is_periodic is set to 1 in all directions, the linear solver returns the right answer. The problem showed up when I tried the symmetry BC in the z-direction (reflecting odd and even for different component appropriately) and it didn't work. Then, I used my ext_dir to set symmetry BC, that didn't work either, and then I thought to go one more step back and use the ext_dir and enforce periodicity (given that I had the exact solution) and see if that works or not, which it didn't either. My apologies in advance, I know it is a bit confusing situation. I'm very confused myself too :D

**Reply by WeiqunZhang** [reply]
Think for the details.  I am not sure why that didn't work, especially when multiple cells in z-direction were used.  However, we do have some support for solving 2D problems in 3D codes by using `LPInfo::setHiddenDirection`.  Here is an example.  https://github.com/WeiqunZhang/amrex-devtests/blob/main/solve_2d_in_3d/MyTest.cpp#L35.

**Reply by alahyari** [reply]
@WeiqunZhang Thank you for your reply. I'm not very familiar with what is under the hood of the AMReX Poisson solve, so I truly appreciate your input. I check the boundary values as well as the RHS values for both scenarios (AMReX periodic and ext_dir periodic) right before the solve call and they are identical, but the solutions after the solve are not the same (AMReX periodic returns the right answer and ext_dir periodic is wrong). So I'm trying to understand what part(s) of the linear solver is possibly not set up right because of which the ext_dir gives a wrong answer. Supposedly, it is a part that is automatically taken care of or overwritten when I use AMReX periodicity. Could you possibly think of any potential cause? 

Thank you very much in advance.

**Reply by WeiqunZhang** [reply]
The only thing I can think of is that the function for handling Dirichlet will fail if there is only one cell in the z-direction.  But you said using multiple cells still gave a wrong answer.  So I don't have any ideas.

It's possible there is something wrong in your setup or in amrex.  However, why do you need to find out, if it only fails in the special setup that you already know the solution.  That's a case you are never going to have in production runs.
